fix search for helper articles

MuskanPaliwal opened the pull request:
fixes #853 

#### Problem
The helper article search was too basic - only did exact text matching. Users struggled to find articles when they made typos or used different word combinations.

#### Solution
Built a more intelligent search algorithm that:
- Handles typos and variations (like "accont" → "account")
- Ranks results better (exact matches first)
- Supports multi-word searches
- Searches both titles and URLs
- Uses fuzzy matching for common mistakes

#### Changes
- Added `helpArticleSearch.ts` with the new search logic
- Updated editor components to use the improved search
- Added unit tests to ensure it works correctly
- Optimized performance with memoization

#### Result: Users can now find helper articles more easily, even when they don't type perfectly.

 
After video: 
https://www.loom.com/share/3cb86236b24b4235917f55529a2a649f?sid=0080843c-d6f8-448b-82de-fc1b045b4fd9

Screenshots:

<img width="624" height="302" alt="Screenshot 2025-08-06 at 1 24 18 PM" src="https://github.com/user-attachments/assets/5fbd03f6-5cee-42ae-a929-ca0dcdb301fe" />

<img width="569" height="318" alt="Screenshot 2025-08-06 at 12 58 23 PM" src="https://github.com/user-attachments/assets/3ad5e9e0-6e88-487c-a13e-b5c3ef8e71c9" />


All tests passing locally

<img width="924" height="425" alt="Screenshot 2025-08-06 at 1 13 19 PM" src="https://github.com/user-attachments/assets/ceb7b92e-0003-46b4-a9b0-7faafbee46d6" />



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Improved help-article search with fuzzy matching, relevance-based ranking and a consistent top-10 results limit for better suggestions when mentioning articles.
* **Bug Fixes**
  * More accurate filtering and ordering of suggestions, including better handling of partial, multi-word, and case-insensitive queries; deterministic alphabetical fallback for empty queries.
* **Tests**
  * Added unit and end-to-end tests covering ranking, filtering, popover behavior, insertion, and closing.
* **Chores**
  * Added a runtime dependency for fuzzy search.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MuskanPaliwal commented:
> An e2e spec would be great!

Sure let me get back on this

MuskanPaliwal commented:
e2e tests for the `helpArticleSearch` feature. All tests passing locally.

Screenshot for reference:
<img width="947" height="248" alt="Screenshot 2025-08-08 at 1 28 40 PM" src="https://github.com/user-attachments/assets/bb49cade-6107-45f9-8df8-a6c6ea999cbf" />

@slavingia , would appreciate your review on this.

MuskanPaliwal commented:
> Sorry for the delay! Is there a library we can use for this, e.g. Fuse.js? (there isn't a huge amount of logic but still feels like something we shouldn't need to worry about maintaining ourselves)

Yes, I had used Fuse only earlier but then instead of adding one more library I decided to go ahead with building our own logic.

binary-koan commented:
I would rather add the library then if it works well enough (adding one more to our >100 dependencies won't make too much difference 😅)

binary-koan commented on components/tiptap/editor.tsx:
```diff
@@ -360,7 +361,9 @@ const TipTapEditor = ({
   };
 
-  const filteredArticles = helpArticles.filter((a) => a.title.toLowerCase().includes(getMentionQuery().toLowerCase()));
+  const filteredArticles = useMemo(() => {
+    return searchHelpArticles(helpArticles, getMentionQuery(), 10);
+  }, [helpArticles, getMentionQuery()]);
 
```
^ I think this is worth doing

MuskanPaliwal commented:
@binary-koan , addressed all your comments on the review . Here's a screenshot of the e2e tests being passed successfully:
<img width="702" height="503" alt="Screenshot 2025-08-13 at 3 10 44 PM" src="https://github.com/user-attachments/assets/351f5186-4506-4151-b578-929e916cb67b" />

Kindly review the PR again.

MuskanPaliwal commented:
> LGTM, sorry about the wait!

Thanks!
